geocoder_tester/base.py

```diff
@@ -133,13 +193,10 @@ def compare_values(get, expected):
 
 def assert_search(query, expected, limit=1,
                   comment=None, lang=None, center=None):
-    params = {"q": query, "limit": limit}
-    if lang:
-        params['lang'] = lang
-    if center:
-        params['lat'] = center[0]
-        params['lon'] = center[1]
-    results = search(**params)
+    api = API_TYPES[CONFIG['API_TYPE']]()
+    params = api.search_params(query, limit, lang, center)
+
+    results = search(api.search_url(), **params)
```

Comment:
Thinking about:

- moving the API initialization inside the `search`, as this is the public API of `geocode_tester`, and one should not have to know how to instantiate the API class to issue a search call (see for example first test in `iledefrance/test_addresses.py`)
- moving the http part inside the `GenericAPI.__call__`

Something like this:

```
class GenericAPI:

    # That is basically the search
    __cal__(self, **params):
        params = self.search_params(params)
        url = self.search_url()
        r = http.get(url, params=params, headers=…)
        …

# Then
def search(**params):
    api = API_TYPES[CONFIG["API_TYPE"]]()
    results = api(query=query, limit=limit, lang=lang, center=center)

# Then inside a test or in `assert_search`:
results = search(q="xxx", center=[1, 2])
```

Thoughts ?

Reply:
I wasn't aware that `search` is a public interface. I've given it a try. This is a bit tricky because currently `search` expects raw HTTP query parameters while `assert_query` expects a a more abstract representation of the query. The former is naturally not compatible with this PR, so I've changed `search` to expect abstract parameters and the ilethefrance test to supply them.

The reason that parameter translation was so explicitly done in `assert_search` is that SearchException expects the translated parameter list and I thought that this is a good thing to keep for debugging purposes. I've changed SearchException now, so that it looks up the parameters again. It's not the prettiest but okay.

(For context, I have an additional constraint here: the GenericAPI class will get three more functions `reverse`, `reverse_params` and `reverse_url` in the next PR. That's why I didn't go for `__call__` and also dropped the idea of params being in the state of GenericAPI.)